### PR TITLE
storage: add assertions that detect HardState proto change

### LINKED ISSUE #17146: storage: improve assertions that prevent HardState proto change
A picture says about as much as a hundred words. From https://reviewable.io/reviews/cockroachdb/cockroach/17051#-KpKIEcA5LZyVrn41G6M

image

### PR DESCRIPTION
**Pull request by pydima**
Closes #17146.

### CODE REVIEW

**Comment by tschottdorf** [comment]
Thanks @pydima!

I think you can simplify this by just adding a test here:

```go
func TestHardStateKnownFields(t *testing.T) {
  populatedHardState := belowRaftGoldenProtos[reflect.TypeOf(&raftpb.HardState{})].populatedConstructor(zeroSeededRand).(*raftpb.HardState)
  // Use reflection to walk all the values and explode if you find a zero value (except for `XXX_unrecognized`)
  // If one is found, Fatal with a hint to update populatedConstructor for *HardState
}
```

**Comment by petermattis** [comment]
If the goal is to detect when there are additional fields added to `raftpb.HardState`, another approach would be to make a parallel structure that is convertible:

```
type expectedHardState struct {
	Term             uint64
	Vote             uint64
	Commit           uint64
	XXX_unrecognized []byte
}

var _ = expectedHardState(raftpb.HardState{})  // This will not compile if fields are added to raftpb.HardState
```

**Comment by tschottdorf** [comment]
That's much easier. Good idea!

**Comment by pydima** [comment]
Thanks, that's really much simpler. I've updated PR.

File: pkg/storage/below_raft_protos_test.go [hunk 1]
```diff
@@ -74,8 +85,6 @@ var belowRaftGoldenProtos = map[reflect.Type]fixture{
 	reflect.TypeOf(&raftpb.HardState{}): {
 		populatedConstructor: func(r *rand.Rand) protoutil.Message {
 			n := r.Uint64()
-			// NB: this would rot if HardState ever picked up more (relevant)
-			// fields, but that's unlikely.
 			return &raftpb.HardState{
```
Comment (tschottdorf) [inline pkg/storage/below_raft_protos_test.go]: Try just adding `XXX_unrecognized: nil,` here and see if that allows you to remove the linter warning.

Comment (pydima) [reply pkg/storage/below_raft_protos_test.go]: yep, you are right, thanks! fixed that.

**Comment by tschottdorf** [comment]
Thanks! I made some small edits to your branch (hope that's ok, just wanted to save you the hassle) and will merge on green.

**Pull request merged**